plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -696,6 +725,22 @@ enterComputeCek = computeCek
     -- FIXME: add rule for VBuiltin once it's in the specification.
     returnCek (FrameApplyFun fun ctx) arg =
         applyEvaluate ctx fun arg
+    -- s , [_ V1 .. Vn] ◅ lam x (M,ρ)  ↦  s , [_ V2 .. Vn]; ρ [ x  ↦  V1 ] ▻ M
+    returnCek (FrameApplyValues args ctx) fun = case args of
+        (arg:rest) -> applyEvaluate (FrameApplyValues rest ctx) fun arg
+        _          -> returnCek ctx fun
+    -- s , constr I V0 ... Vj-1 _ (Tj+1 ... Tn, ρ) ◅ Vj  ↦  s , constr i V0 ... Vj _ (Tj+2... Tn, ρ)  ; ρ ▻ Tj+1
+    returnCek (FrameConstr env i todo done ctx) e = do
+        let done' = done `DList.snoc` e
+        case todo of
+          (next : todo') -> computeCek (FrameConstr env i todo' done' ctx) env next
+          _              -> returnCek ctx $ VConstr i (toList done')
+    -- s , case _ (C0 ... CN, ρ) ◅ constr i V1 .. Vm  ↦  s , [_ V1 ... Vm] ; ρ ▻ Ci
+    returnCek (FrameCases env cs ctx) e = case e of
+        (VConstr i args) -> case cs ^? ix i of
```

Comment:
Does it make sense to use `Vector` for `cs` for faster indexing?

Reply:
Ah yes. This needs a Note. In short, I spent a _long_ time trying out various accumulator types here in search of more performance. I started out with `Vector`/`MutableVector` as the "obvious" choice, at @effectfully 's prompting I tried `Seq` and also `List`/`DList`. To my great surprise, there was no real difference, and so I've gone with `List`/`DList` as it's the simplest.